components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -54,3 +176,327 @@ impl ReadableByteStreamControllerMethods for ReadableByteStreamController {
+            Box::new(Self { controller })
+        }
+    }
+
+    impl Callback for ResolveHandler {
+        fn callback(&self, cx: SafeJSContext, _v: SafeHandleValue, _realm: InRealm) {
+            // Step 11.1
+            self.controller.started.set(true);
+            // Step 11.2
+            assert!(!self.controller.pulling.get());
+            // Step 11.3
+            assert!(!self.controller.pull_again.get());
+            // Step 11.4
+            assert!(readable_byte_stream_controller_call_pull_if_needed(
+                cx,
+                self.controller.clone()
+            )
+            .is_ok());
+        }
+    }
+
+    Ok(())
+}
+
+/// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-call-pull-if-needed>
+fn readable_byte_stream_controller_call_pull_if_needed(
+    cx: SafeJSContext,
+    controller: DomRoot<ReadableByteStreamController>,
+) -> Fallible<()> {
+    // Step 1 & 2
+    if controller.should_call_pull() {
+        // Step 3
+        if controller.pulling.get() {
+            controller.pull_again.set(true);
+        } else {
+            // Step 4
+            assert!(!controller.pull_again.get());
+            // Step 5
+            controller.pulling.set(true);
+            // Step 6
+            let pull_promise = controller.algorithms.borrow().pull(
+                cx,
+                ReadableStreamController::ReadableByteStreamController(controller.clone()),
+            )?;
+            let global = &*controller.global();
+            let realm = enter_realm(&*global);
+            let comp = InRealm::Entered(&realm);
+            pull_promise.append_native_handler(
+                &PromiseNativeHandler::new(
+                    global,
+                    Some(ResolveHandler::new(controller.clone())),
+                    Some(RejectHandler::new(controller)),
+                ),
+                comp,
+            );
+        }
+    }
+
+    #[derive(JSTraceable, MallocSizeOf)]
+    struct ResolveHandler {
+        controller: DomRoot<ReadableByteStreamController>,
+    }
+
+    impl ResolveHandler {
+        pub fn new(controller: DomRoot<ReadableByteStreamController>) -> Box<dyn Callback> {
+            Box::new(Self { controller })
+        }
+    }
+
+    impl Callback for ResolveHandler {
+        fn callback(&self, cx: SafeJSContext, _v: SafeHandleValue, _realm: InRealm) {
+            // Step 7.1
+            self.controller.pulling.set(false);
+            // Step 7.2
+            if self.controller.pull_again.get() {
+                self.controller.pull_again.set(false);
+                assert!(readable_byte_stream_controller_call_pull_if_needed(
+                    cx,
+                    self.controller.clone()
+                )
+                .is_ok());
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Algorithms for [setup_readable_stream_default_controller_from_underlying_source]
+#[derive(JSTraceable, MallocSizeOf)]
+pub enum ControllerAlgorithms {
```

Comment:
The controllers have several internal slots that are "Algorithms". I found it difficult to use generic in the controller's field, because they can be changed later. Perhaps it's better to use `dyn triat`?